enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,232 @@
+
+#### Story 2:
+
+As an openshift developer I’m assigned a release blocker BZ regarding upstream implementation. Proving the value for merging this and making the bug a release blocker upstream requires careful communication with upstream maintainers, and explaining the value to the people whom are not necessarily involved into internal details, so their opinion could differ and delay merging an important fix.
+
+#### Story 3:
+
+We desire to extend our team responsibilities on upstream cloud providers in the future and gain weight in promoting features into upstream. Getting such for kubernetes repository is harder for us and maintainers, as this would mean giving weight for approving and merging features for the parts of the kubernetes project, not necessarily laying under our responsibilities and scope of knowledge.
+
+#### Story 4:
+
+We’d like to discuss technical details related to a specific cloud in a SIG meeting with people who are also involved into development in this domain, and that way gain useful insights into the cloud infrastructure, improve the overall quality of our features, stay on top of the new features, and improve the relations with maintainers outside of our company, which nevertheless share with us a common goal.
+
+
+### Implementation Details
+
+The `external` `cloud-controller-manager` implementation will be hosted within openshift repository in https://github.com/openshift/cluster-cloud-controller-manager-operator. That repository will manage `CCM` provisioning for metioned cloud providers + `Openstack` out-of-tree provider implementation described in the [document](https://github.com/openshift/enhancements/tree/master/enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md)
+
+`Cloud-controller` will be provisioned in the `openshift-cloud-controller-manager` namespace.
+
+Preparation steps, common for all providers:
+
+- Add support for recognition of the `external` option in [library-go](https://github.com/openshift/library-go/blob/master/pkg/operator/configobserver/cloudprovider/observe_cloudprovider.go) which will disable config parsing and add support for `external` field in the `Infrastructure` resource.
+- Port the templates of `cloud-controller` from upstream repositories into proposed https://github.com/openshift/cluster-cloud-controller-manager-operator repository. Those templates will belong to the newly created `openshift-cloud-controller-manager` namespace and be managed with an operator.
```

Comment:
Which templates is this referring to,  can we please elaborate this a bit?

Reply:
Agreed, would be good to link here to the various templates for the providers so that we have an understanding of what we are actually going to be deploying here